Inventory Detail

keithjgrant opened the pull request:
##### SUMMARY
Add Inventory Detail screen

related #5261 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI


##### ADDITIONAL INFORMATION
This also introduces new components/refactors:

* `VariablesDetail` providing a read-only view of YAML/JSON variables, for use inside a `DetailList`
* `TabbedCardHeader` to normalize the styling of all Cards with tabs in the header
* `UserDateDetail` to display created/modified dates and a link to the user who made the change, for use in a `DetailList`

All relevant screens with these things have been updated to use these components.

keithjgrant committed:
Start inventory detail

* Create VariablesDetail for read-only variables view
* Sketch out InventoryDetail
* Create CardBody and TabbedCardHeader for common custom styling

keithjgrant committed:
finish InventoryDetail

keithjgrant committed:
add tests for VariablesDetail & InventoryDetail

keithjgrant committed:
switch all tabbed screens to use TabbedCardHeader

keithjgrant committed:
use VariablesDetail for displaying variables field in details views

jakemcdermott commented on awx/ui_next/src/components/DetailList/UserDateDetail.jsx:
```diff
@@ -0,0 +1,31 @@
+  return (
+    <Detail
+      label={label}
+      value={
```
The code this replaces has a translated value, so we may want to also include that here. As I recall, the translated value should (ideally) be a single string.

See: https://github.com/ansible/awx/pull/4915#discussion_r331644299 for details.

keithjgrant replied:
Good catch. Fixed

jakemcdermott commented on awx/ui_next/src/screens/Host/HostDetail/HostDetail.jsx:
```diff
@@ -66,23 +41,20 @@ function HostDetail({ host, i18n }) {
+        <UserDateDetail
+          label={i18n._(t`Last Modified`)}
+          date={modified}
+          user={summary_fields.modified_by}
```
For a given resource in awx, it is not guaranteed that `modified_by`, `created_by` will be defined on `summary_fields` so you'll want to check that it is defined before referencing it to avoid a potential error.

keithjgrant replied:
The `UserDateDetail` handles null/undefined values gracefully, so this works out

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryDetail/InventoryDetail.jsx:
```diff
@@ -1,10 +1,84 @@
+          fullWidth
+          label={i18n._(t`Instance Groups`)}
+          value={
+            isLoading ? (
```
Individual loading states for specific fields is a new UX that we don't see on the other views. I have _some_ reservations about going down this path. How would this behave for lots of different fields that have to load asynchronously, like we see on job templates? Error handling?

keithjgrant replied:
Good call. I'll make this more in line with existing Details pages...

jakemcdermott commented on awx/ui_next/src/components/Card/TabbedCardHeader.js:
```diff
@@ -0,0 +1,12 @@
+import styled from 'styled-components';
+import { CardHeader } from '@patternfly/react-core';
+
+const TabbedCardHeader = styled(CardHeader)`
```
Nice. :+1: This was a good (re) addition and cuts out some duplication.

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryDetail/InventoryDetail.jsx:
```diff
@@ -1,10 +1,84 @@
+  useEffect(() => {
+    (async () => {
+      setIsLoading(true);
+      const { data } = await InventoriesAPI.readInstanceGroups(inventory.id);
```
Error handling is needed for this request.

keithjgrant committed:
use UserDateDetail in OrganizationDetail

keithjgrant committed:
InventoryDetail: handle content loading state & errors better

keithjgrant committed:
Fix UserDateDetail translation

Add UserDateDetail to Org detail & InventoryGroupDetail
Add VariablesDetail to InventoryGroupDetail

keithjgrant committed:
switch to our CardBody component on all screens, for consistent padding/spacing

keithjgrant committed:
Add Inventory detail edit button

keithjgrant committed:
add delete button to InventoryDetail

keithjgrant committed:
update tests for detail view changes

keithjgrant committed:
de-lint
